internal/mining/mining.go

```diff
@@ -576,20 +606,110 @@ func createCoinbaseTx(subsidyCache *standalone.SubsidyCache, coinbaseScript []by
+	// is included in the stake tree instead.
+	var txVersion = uint16(1)
+	var treasuryOutput *wire.TxOut
+	var treasurySubsidy int64
+	if !isTreasuryEnabled {
+		if params.BlockTaxProportion > 0 {
+			// Create the treasury output with the correct subsidy and public
+			// key script for the organization associated with the treasury.
+			treasurySubsidy = subsidyCache.CalcTreasurySubsidy(nextBlockHeight,
+				voters)
+			treasuryOutput = &wire.TxOut{
+				Value:    treasurySubsidy,
+				PkScript: params.OrganizationPkScript,
+			}
+		} else {
+			// Treasury disabled.
+			treasuryOutput = &wire.TxOut{
+				Value:    0,
+				PkScript: opTrueScript,
+			}
+		}
+	} else {
+		// Set the transaction version to the new version required by the
+		// decentralized treasury agenda.
+		txVersion = wire.TxVersionTreasury
+	}
+
+	// Create a coinbase with expected inputs and outputs.
+	//
+	// Inputs:
+	//  - A single input with input value set to the total payout amount.
+	//
+	// Outputs:
+	//  - Potential treasury output prior to the decentralized treasury agenda
+	//  - Output that includes the block height and potential extra nonce used
+	//    to ensure a unique hash
+	//  - Output that pays the work subsidy to the miner
 	workSubsidy := subsidyCache.CalcWorkSubsidy(nextBlockHeight, voters)
+	tx := wire.NewMsgTx()
+	tx.Version = txVersion
+	tx.AddTxIn(coinbaseInput)
+	tx.TxIn[0].ValueIn = workSubsidy + treasurySubsidy
+	if treasuryOutput != nil {
+		tx.AddTxOut(treasuryOutput)
+	}
+	tx.AddTxOut(&wire.TxOut{
+		Value:    0,
+		PkScript: opReturnPkScript,
+	})
+	tx.AddTxOut(&wire.TxOut{
+		Value:    workSubsidy,
+		PkScript: workSubsidyScript,
+	})
+	return dcrutil.NewTx(tx), nil
+}
+
+// createTreasuryBaseTx returns a coinbase transaction paying an appropriate subsidy
+// based on the passed block height to the treasury.
+func createTreasuryBaseTx(subsidyCache *standalone.SubsidyCache, opReturnPkScript []byte, nextBlockHeight int64, voters uint16, params *chaincfg.Params) (*dcrutil.Tx, error) {
+	tx := wire.NewMsgTx()
+	// All treasury base transactions require TxVersionTreasury.
+	tx.Version = wire.TxVersionTreasury
+
+	tx.AddTxIn(&wire.TxIn{
+		// Stakebase transactions have no inputs, so previous outpoint
+		// is zero hash and max index.
+		PreviousOutPoint: *wire.NewOutPoint(&chainhash.Hash{},
+			wire.MaxPrevOutIndex, wire.TxTreeRegular),
+		Sequence:        wire.MaxTxInSequenceNum,
+		BlockHeight:     wire.NullBlockHeight,
+		BlockIndex:      wire.NullBlockIndex,
+		SignatureScript: nil, // Must be nil by consensus.
+	})
+
+	// Create a stakebase with correct block subsidy and extranonce.
 	treasurySubsidy := subsidyCache.CalcTreasurySubsidy(nextBlockHeight, voters)
 
 	// Treasury output.
 	if params.BlockTaxProportion > 0 {
 		tx.AddTxOut(&wire.TxOut{
 			Value:    treasurySubsidy,
-			PkScript: params.OrganizationPkScript,
+			PkScript: []byte{txscript.OP_TADD},
 		})
 	} else {
```

Comment:
This is a problem that we'll need to discuss how to handle.  The current detection logic for a treasurybase transaction will not allow the output to be `OP_TRUE` as this code does in the case the config parameter is 0 (meaning no treasury payout).  That means it would create treasurybase transactions that would be rejected thus halting the chain.

So, I think we have 2 choices here to keep the ability to handle 0 treasury proportion configs properly:

1) We set the `Value` field of the `OP_TADD` output to 0 and deal with the fallout on the accounting side -- I'm not zero amounts will work properly without digging into the relevant code, but even if they do we probably wouldn't want a bunch of zero entries in the journals); or
2) We allow the output to be omitted, which I suspect would not only complicate the detection logic, it would also complicate the accounting aspects

The other alternative, which I think might be a better choice, is to rip out the ability for it to be 0 throughout.  We need to verify, but I'm fairly certain that wouldn't be a consensus change since it has never been zero and therefore explicitly prohibiting it would have no effect.